[Do-not-merge] Kafka authentication with user/pass

annanay25 opened the pull request:
## Description
Adds authentication to connect to Kafka through the Sarama library. 
Yet to confirm exhaustive list of config attributes required.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [ ] All new and existing tests passed.

Commit:
Kafka authentication with user/pass

nitisht committed:
Add S3 gateway documentation (#6165)

Fixes #4830

Commit:
Cleaner config store

vadmeste committed:
Disable splitting lines in pretty error messages (#6171)

In a small window, UI error tries to split lines for an eye candy
error message. However, since we show some docs.minio.io links in some
error messages, these links are actually broken and not easily selected
in a X terminal. This PR changes the behavior and won't split lines
anymore.

vadmeste committed:
Add new console/http loggers (#6066)

- Add console target logging, enabled by default.
- Add http target logging, which supports an endpoint
  with basic authentication (username/password are passed
  in the endpoint url itself)
- HTTP target logging is asynchronous and some logs can be
  dropped if channel buffer (10000) is full

balamurugana commented on pkg/event/target/kafka.go:
```diff
@@ -32,6 +32,21 @@ type KafkaArgs struct {
+				ClientAuth         int  `json:"clientauth"`
+			} `json:"tlsconfig"`
+		} `json:"tls"`
+	} `json:"net"`
```
1. Nesting of SASL/TLS configuration is not good idea.  It will lead to complicated configuration file.
2. Nested TLS.Config is a bad idea.
3. Is Kafka used with SASL and TLS independently?

balamurugana replied:
Keep simplified configuration change like below

```golang
// KafkaArgs - Kafka target arguments.
type KafkaArgs struct {
	Enable  bool        `json:"enable"`
	Brokers []xnet.Host `json:"brokers"`
	Topic   string      `json:"topic"`
	TLS     struct {
		Enable     bool               `json:"enable"`
		SkipVerify bool               `json:"skipVerify"`
		ClientAuth tls.ClientAuthType `json:"clientAuth"`
	} `json:"tls"`
	SASL struct {
		Enable    bool   `json:"enable"`
		Username  string `json:"username"`
		Password  string `json:"password"`
		Handshake bool   `json:"handshake"`
	} `json:"sasl"`
}
```

annanay25 replied:
+1 @balamurugana 

However, are we planning to switch the client library to segmentio/kafka-go? The library does not have authentication yet.

harshavardhana replied:
If it's not then we don't need to migrate, we can migrate when it implements authentication. @annanay25 But they implement TLS based encryption https://github.com/segmentio/kafka-go/issues/109#issuecomment-403086031

balamurugana commented on pkg/event/target/kafka.go:
```diff
@@ -90,9 +105,22 @@ func (target *KafkaTarget) Close() error {
+                InsecureSkipVerify: KafkaArgs.Net.TLS.Config.InsecureSkipVerify,
+                ClientAuth: KafkaArgs.Net.TLS.Config.ClientAuth,
+        }
+        config.Net.TLS.Config = tlsConfig
```
Any configuration change requires version bump and migration code from old to new configuration.  You would need to do that in https://github.com/minio/minio/blob/master/cmd/config-versions.go, https://github.com/minio/minio/blob/master/cmd/config-current.go and https://github.com/minio/minio/blob/master/cmd/config-migrate.go

kannappanr committed:
Do not trace HTTP body of PostPolicyBuckethandler (#6177)

harshavardhana committed:
Remove unused functions and constants (#6175)

kannappanr committed:
Remove ppc64le support from .travis.yml (#6180)

Travis CI for ppc64le stays in queued state.
Removing the support for now.

kannappanr committed:
Log x-amz-request-id as log and XML error response (#6173)

Currently, requestid field in logEntry is not populated, as the
requestid field gets set at the very end.
It is now set before regular handler functions. This is also
useful in setting it as part of the XML error response.

Travis build for ppc64le has been quite inconsistent and stays queued
for most of the time. Removing this build as part of Travis.yml for
the time being.

harshavardhana committed:
Do not use parallel deletes to avoid random I/O (#6178)

The current code for deleting 1000 objects simultaneously
causes significant random I/O, which on slower drives
leads to servers disconnecting in a distributed setup.

Simplify this by serially deleting and reducing the
chattiness of this operation.

kannappanr committed:
Do not set Key and BucketName in ErrorResponse, if empty (#6174)

harshavardhana committed:
S3 gateway signature probe use a unique bucket (#6190)

This fixes an issue because someone is
using `probe-bucket-sign` bucket name in
region 'eu-central-1'

minio-trusted committed:
Update yaml files to latest version RELEASE.2018-07-23T18-34-49Z

harshavardhana committed:
Fix healthcheck handler to check errors in local disks only (#6184)

Healthcheck handler in current implementation was
performing ListBuckets() to check for liveness of Minio
service. ListBuckets() implementation on the other hand
doesn't do quorum based listing and if one of the disks
returned error, an I/O error it would be lead to kubernetes
taking the minio pod down prematurely even if the disk
is not local to that minio server.

The reason is ListBuckets() call cannot be trusted to
provide us the valid information that we need, Minio is a
clustered application which is designed to handle disk
failures. Error on one of the disks doesn't mean the pod
should become fully non-operational.

This PR attempts to fix this by only checking for alive
disks which are local to each setup and also by simply
performing a Stat() operation, if the Stat() returned
error on all disks local to a particular server then
we can let kubernetes safely take it down, until then
we should be operational.

kaankabalak committed:
Replace pointer with default cursor on filename hover (#6193)

kannappanr committed:
Fix make test failure (#6185)

Fix shadowing errors.

harshavardhana committed:
Remove references to MINIO_ENDPOINTS (#6200)

MINIO_ENDPOINTS is a special case scenario
we don't need to document it for now.

harshavardhana committed:
Remove error returned when part sizes are un-equal (#6183)

Since implementing `pwrite` like implementation would
require a more complex code than background append
implementation, it is better to keep the current code
as is and not implement `pwrite` based functionality.

Closes #4881

Commit:
crypto: add support for parsing/creating SSE-C/SSE-S3 metadata (#6169)

* crypto: add support for parsing SSE-C/SSE-S3 metadata

This commit adds support for detecting and parsing
SSE-C/SSE-S3 object metadata. With the `IsEncrypted`
functions it is possible to determine whether an object
seems to be encrypted. With the `ParseMetadata` functions
it is possible to validate such metadata and extract the
SSE-C/SSE-S3 related values.

It also fixes some naming issues.

* crypto: add functions for creating SSE object metadata

This commit adds functions for creating SSE-S3 and
SSE-C metadata. It also adds a `CreateMultipartMetadata`
for creating multipart metadata.

For all functions unit tests are included.

Commit:
Simplified KafkaArgs struct, added sample config

harshavardhana committed:
Make sure to handle FaultyDisks in listing ops (#6204)

Continuing from PR 157ed65c352e40c71fe6ab91738321d95bd19b34

Our posix.go implementation did not handle  I/O errors
properly on the disks, this led to situations where
top-level callers such as ListObjects might return early
without even verifying all the available disks.

This commit tries to address this in Kubernetes, drbd/nbd based
persistent volumes which can disconnect under load and
result in the situations with disks return I/O errors.

This commit also simplifies listing operation, listing
never returns any error. We can avoid this since we pretty
much ignore most of the errors anyways. When objects are
accessed directly we return proper errors.

donatello committed:
Fix type of bitrot mismatch error (#6205)

The error type `hashMismatchError` is lost when the error is received
from a remote disk.

Fixes #6201

harshavardhana committed:
Remove double RLocks for GetBucketInfo (#6209)

minio-trusted committed:
Update yaml files to latest version RELEASE.2018-07-31T02-11-47Z

kannappanr committed:
Return 503 instead of 404 if more than half of disks are not found (#6207)

Fixes #6163

nitisht committed:
Fix healthcheck handler to verify gateway backend liveness (#6218)

Fixes #6217

nitisht committed:
Fix Kubernetes TLS doc to avoid creating CAs dir on read only mount (#6214)

Commit:
crypto: implement Stringer for S3 and SSEC (#6216)

This commit adds a `fmt.Stringer` implementation for
SSE-S3 and SSE-C. The string representation is the
domain used for object key sealing.
See: `ObjectKey.Seal(...)` and `ObjectKey.Unseal(...)`

aarushiarya committed:
Appropriate error message on unsuccessful update. (#6203)

Commit:
crypto: add `GenerateIV` from random IV generation (#6215)

This commit adds a `GenerateIV` function to simplify
the generation of random IVs.

It adds some unit tests for `GenerateIV` in key_test.go

harshavardhana committed:
Handle POST object upload without filename param (#6221)

POST mime/multipart upload style can have filename value optional
which leads to implementation issues in Go releases in their
standard mime/multipart library.

When `filename` doesn't exist Go doesn't update `form.File` which
we rely on to extract the incoming file data, strangely when `filename`
is not specified this data is buffered in memory and is now part of
`form.Value` instead of `form.File` which creates an inconsistent
behavior.

This PR tries to fix this in our code for the time being, but ideal PR
would be to fix the upstream mime/multipart library to handle the
above situation consistently.

poornas committed:
check for syscall errors in posix-errors helper functions (#6232)

annanay25 committed:
Added config and tests

annanay25 committed:
Added config and tests

annanay25 committed:
Rebased

annanay25 committed:
Rebasing

annanay25 committed:
Fixing tests

kannappanr commented on cmd/config-migrate.go:
```diff
@@ -2355,3 +2360,123 @@ func migrateV26ToV27() error {
+		return nil
+	}
+
+	// Copy over fields from V26 into V27 config struct
```
Should be `// Copy over fields from V27 into V28 config struct`

kannappanr commented on cmd/config-versions.go:
```diff
@@ -722,3 +722,37 @@ type serverConfigV27 struct {
 	Logger loggerConfig `json:"logger"`
 }
+
+// serverConfigV27 is backward compatible with version '26'.
```
Do we need info about version27 here?

kannappanr commented on docs/config/config.sample.json:
```diff
@@ -91,7 +91,17 @@
                 "enable": true,
                 "brokers": ["localhost:9092"],
-                "topic": "bucketevents"
+                "topic": "bucketevents",
```
The version needs to be changed to `28`

annanay25 committed:
Fixes config version
